dockerfile_parse/util.py

```diff
@@ -188,3 +188,108 @@ def remove_nonescaped_quotes(string):
     string = remove_quotes(string)
     string = string.replace('\ ', ' ')
     return string.replace("\\s", "'").replace('\\d', '"')  # restore
+
+
+def extract_labels_or_envs(env_replace, envs, instruction_value):
+    shlex_splits_raw = shlex_split(instruction_value,
+                                   env_replace=env_replace, envs=envs)
+
+    key_val_list = []
+    if '=' not in shlex_splits_raw[0]:  # LABEL/ENV name value
+        # split it to first (name) and the rest (value)
+        key_val = instruction_value.split(None, 1)
+        key = strip_quotes(key_val[0])
+        try:
+            val = key_val[1]
+        except IndexError:
+            val = ''
+
+        if env_replace:
+            val = EnvSubst(val, envs).substitute()
+        val = remove_nonescaped_quotes(val)
+
+        key_val_list.append((key, val))
+
+    else:  # LABEL/ENV "name"="value"
+        shlex_splits = []
```

Comment:
iiuc, `shlex_splits` is a list of strings in the form of `'key=value'`. It seems that this can be better handled by using a different data structure, either a dict or a list of tuples. Something like:

```python
shlex_splits = {}
current_key = None
for part in shlex_splits_raw:
  if current_key:
    shlex_splits[current_key] = part
    current_key = None
    continue
  if part != '=':
    current_key = part

key_val_list.extend(shlex_splits.items())
```